lib/posix/pthread.c

```diff
@@ -151,69 +286,85 @@ static void zephyr_thread_wrapper(void *arg1, void *arg2, void *arg3)
-	    || (attr->stack == NULL) || (attr->stacksize == 0)) {
+	if (!pthread_attr_is_valid((const struct pthread_attr *)_attr)) {
 		return EINVAL;
 	}
 
-	key = k_spin_lock(&pthread_pool_lock);
-	for (pthread_num = 0;
-	    pthread_num < CONFIG_MAX_PTHREAD_COUNT; pthread_num++) {
-		thread = &posix_thread_pool[pthread_num];
-		if (thread->state == PTHREAD_EXITED || thread->state == PTHREAD_TERMINATED) {
-			thread->state = PTHREAD_JOINABLE;
-			break;
+	/* copy the attribute in to the */
+	(void)memcpy(&attr, _attr, sizeof(attr));
+
+	key = k_spin_lock(&q_lock);
+	/* recycle old nodes if they are available */
+	do {
+		struct posix_thread *safe_t;
+
+		SYS_DLIST_FOR_EACH_CONTAINER_SAFE(&done_q, t, safe_t, q_node) {
+			if (t->detachstate == PTHREAD_CREATE_JOINABLE) {
+				/* thread has not been joined yet */
+				continue;
+			}
+			sys_dlist_remove(&t->q_node);
+			sys_dlist_append(&ready_q, &t->q_node);
+			t->qid = POSIX_THREAD_READY_Q;
 		}
+	} while (false);
+
+	/* check if a thread is available to spawn */
+	t = NULL;
+	if (!sys_dlist_is_empty(&ready_q)) {
+		t = CONTAINER_OF(sys_dlist_get(&ready_q), struct posix_thread, q_node);
+		t->qid = POSIX_THREAD_NONE_Q;
 	}
-	k_spin_unlock(&pthread_pool_lock, key);
+	k_spin_unlock(&q_lock, key);
 
-	if (pthread_num >= CONFIG_MAX_PTHREAD_COUNT) {
+	if (t == NULL) {
+		/* no threads are ready */
 		return EAGAIN;
 	}
 
-	rv = pthread_mutex_init(&thread->state_lock, NULL);
-	if (rv != 0) {
-		key = k_spin_lock(&pthread_pool_lock);
-		thread->state = PTHREAD_EXITED;
-		k_spin_unlock(&pthread_pool_lock, key);
-		return rv;
+	err = pthread_barrier_init(&barrier, NULL, 2);
+	if (err != 0) {
+		key = k_spin_lock(&q_lock);
+		sys_dlist_append(&ready_q, &t->q_node);
+		t->qid = POSIX_THREAD_READY_Q;
+		k_spin_unlock(&q_lock, key);
+		/* can't allocate the barrier to securely spawn a thread */
+		return EAGAIN;
 	}
 
-	prio = posix_to_zephyr_priority(attr->priority, attr->schedpolicy);
+	/* modify thread state atomically */
+	key = k_spin_lock(&t->lock);
+	t->detachstate = attr.detachstate;
+	if ((BIT(_PTHREAD_CANCEL_POS) & attr.flags) != 0) {
+		t->cancel_state = PTHREAD_CANCEL_ENABLE;
+	}
+	t->cancel_pending = false;
+	sys_slist_init(&t->key_list);
+	k_spin_unlock(&t->lock, key);
 
-	cancel_key = k_spin_lock(&thread->cancel_lock);
-	thread->cancel_state = (1 << _PTHREAD_CANCEL_POS) & attr->flags;
-	thread->cancel_pending = 0;
-	k_spin_unlock(&thread->cancel_lock, cancel_key);
+	/* spawn the thread and then wait for the spawned thread to cross our barrier */
+	k_thread_create(&t->thread, attr.stack, attr.stacksize, zephyr_thread_wrapper, (void *)arg,
+			UINT_TO_POINTER(barrier), threadroutine,
+			posix_to_zephyr_priority(attr.priority, attr.schedpolicy),
+			(~K_ESSENTIAL & attr.flags), K_MSEC(attr.delayedstart));
 
-	pthread_mutex_lock(&thread->state_lock);
-	thread->state = attr->detachstate;
-	pthread_mutex_unlock(&thread->state_lock);
+	err = pthread_barrier_wait(&barrier);
```

Comment:
Ah, it looks like you're using the barrier to avoid synchronization bugs in the underlying k_thread implementation. If so, it would be good to mention that here -- it seems to me that a "correct" k_threads implementation would not require the thread to have started before being able to manipulate it with further k_thread API calls from the parent thread?

Reply:
`k_thread` is fine (I think? although there may be lingering arch-specific issues).

The thing that we want to synchronize is that the thread should be started and joinable by the time that `pthread_create()` returns.

Technically, as soon as `k_thread_create()` returns here, it should be possible for `k_thread_join()` to be called. The `k_thread` implementation is not the problem. It's actually `pthread_create()` and `pthread_join()` which are problematic - they do not operate in kernel context, and so there is no guarantee if `pthread_create()` will return before `pthread_join()` is called.

The barrier helps to synchronize that event though.

Reply:
I'm confused here (not surprising, this code is the most complex thing I've reviewed in months)...

In order for `pthread_join()` to be callable before `pthread_create()` returns, it would need the thread ID. You've very carefully ensured that the thread id is *not* visible until all of the work in `pthread_create()` has finished. So, it seems to me that you've already guaranteed that `pthread_join()` cannot be called until after `pthread_create()` is done (not quite returned, that's the next statement after the thread id assignment...).

Are you concerned about `pthread_join()` being called with the thread id from the last time the same slot was used?

Reply:
It's not so much that `pthread_join()` can't be called until `pthread_create()` has returned, it's that the new thread itself needs to be in a joinable state before `pthread_create()` returns (and before `pthread_join()` can be called).

That basically reduces to the pthread needing to be in a joinable state before `pthread_join()` can be called.

For pthreads (at least with this fancy new implementation), that means
1. the thread cannot be in `ready_q`
2. it must have been created `PTHREAD_CREATE_JOINABLE` (of course)

I think that's it..

Reply:
so the thread needs to be out of `ready_q` before `pthread_create()` returns. Is there some reason that couldn't happen? Given the cost of the extra context switches, it seems worth a bit of investigation to see if we can't avoid this synchronization ...

Reply:
`k_thread_create()` is a scheduling point, and there is no guarantee that this or the other would be the next context swapped in. At least, that's my guess. Is that a semi-accurate take, @andyross ?

Reply:
Ok, I think I understand things well enough now. As above, I *think* you could move the `run_q` assignment to this function, but that would require a bunch more thinking and it purely a performance optimization.